lib/Dialect/Arith/Transforms/QuarterWideInt.cpp

```diff
@@ -0,0 +1,516 @@
+    rewriter.replaceOp(op, resultVec);
+    return success();
+  }
+};
+
+struct ConvertArithConstant final
+    : OpConversionPattern<mlir::arith::ConstantOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      mlir::arith::ConstantOp op, OpAdaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    Type oldType = op.getType();
+    auto newType = getTypeConverter()->convertType<RankedTensorType>(oldType);
+
+    if (!newType)
+      return rewriter.notifyMatchFailure(
+          op, llvm::formatv("unsupported type: {0}", op.getType()));
+
+    unsigned newBitWidth = newType.getElementTypeBitWidth();
+    Attribute oldValue = op.getValueAttr();
+
+    if (auto intAttr = dyn_cast<IntegerAttr>(oldValue)) {
+      auto [low, midLow, midHigh, high] =
+          getQuarters(intAttr.getValue(), newBitWidth);
+      auto newAttr =
+          DenseElementsAttr::get(newType, {low, midLow, midHigh, high});
+      rewriter.replaceOpWithNewOp<mlir::arith::ConstantOp>(op, newAttr);
+      return success();
+    }
+
+    if (auto splatAttr = dyn_cast<SplatElementsAttr>(oldValue)) {
+      auto [low, midLow, midHigh, high] =
+          getQuarters(splatAttr.getSplatValue<APInt>(), newBitWidth);
+      int64_t numSplatElems = splatAttr.getNumElements();
+      SmallVector<APInt> values;
+      values.reserve(numSplatElems * 4);
+      for (int64_t i = 0; i < numSplatElems; ++i) {
+        values.push_back(low);
+        values.push_back(midLow);
+        values.push_back(midHigh);
+        values.push_back(high);
+      }
+
+      auto attr = DenseElementsAttr::get(newType, values);
+      rewriter.replaceOpWithNewOp<mlir::arith::ConstantOp>(op, attr);
+      return success();
+    }
+
+    if (auto elemsAttr = dyn_cast<DenseElementsAttr>(oldValue)) {
+      int64_t numElems = elemsAttr.getNumElements();
+      SmallVector<APInt> values;
+      values.reserve(numElems * 4);
+      for (const APInt &origVal : elemsAttr.getValues<APInt>()) {
+        auto [low, midLow, midHigh, high] = getQuarters(origVal, newBitWidth);
+        values.push_back(std::move(low));
+        values.push_back(std::move(midLow));
+        values.push_back(std::move(midHigh));
+        values.push_back(std::move(high));
+      }
+
+      auto attr = DenseElementsAttr::get(newType, values);
+      rewriter.replaceOpWithNewOp<mlir::arith::ConstantOp>(op, attr);
+      return success();
+    }
+
+    return rewriter.notifyMatchFailure(op.getLoc(),
+                                       "unhandled constant attribute");
+  }
+};
+
+struct ConvertExtSI final : OpConversionPattern<mlir::arith::ExtSIOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      mlir::arith::ExtSIOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    Location loc = op->getLoc();
+
+    auto newTy =
+        getTypeConverter()->convertType<RankedTensorType>(op.getType());
```

Comment:
I suppose this is not already on the adaptor for some reason?

Reply:
Now newTy is directly of the type RankedTensorType, while with the adaptor I would need to do an additional conversion. 
for me it's the same, what would you prefer?

Reply:
fine to leave as-is